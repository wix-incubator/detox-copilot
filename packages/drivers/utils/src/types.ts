export type ElementCategory =
  | "button"
  | "link"
  | "input"
  | "list"
  | "table"
  | "header"
  | "semantic";
