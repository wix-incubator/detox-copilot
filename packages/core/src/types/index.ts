export * from "@/types/core";
export * from "@/types/framework";
export * from "@/types/prompt";
export * from "@/types/auto";
export * from "@/types/cache";
export * from "@/types/logger";
