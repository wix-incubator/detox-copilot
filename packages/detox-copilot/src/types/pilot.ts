/**
 * Executed test step record.
 */
export type PreviousStep = {
  /** Step description */
  step: string;
  /** Generated test code */
  code: string;
  /** Step execution result */
  result: any;
};

/** Review section types */
export type PilotReviewSectionType = "ux" | "a11y";

/** Complete pilot review */
export type PilotReview = {
  [key in PilotReviewSectionType]?: PilotReviewSection;
};

/**
 * Single pilot step execution report.
 */
export type PilotStepReport = {
  /** Action plan and reasoning */
  plan: PilotStepPlan;
  /** Optional reviews */
  review?: PilotReview;
  /** Generated code */
  code?: string;
  /** Indicates if the goal was achieved */
  goalAchieved: boolean;
  /** */
  summary?: string;
};

/**
 * Complete pilot execution report.
 */
export type PilotReport = {
  /** Target objective */
  goal: string;
  /** Execution summary */
  summary?: string;
  /** Individual step reports */
  steps: PilotStepReport[];
  /** Final reviews */
  review?: PilotReview;
};

/**
 * Pilot step planning output.
 */
export type PilotStepPlan = {
  /** Planned action */
  action: string;
  /** Reasoning process */
  thoughts: string;
};

/**
 * Review section content.
 */
export type PilotReviewSection = {
  /** Overall assessment */
  summary: string;
  /** Specific observations */
  findings?: string[];
  /** Numerical rating (1-10) */
  score: string;
};

/**
 * Previous pilot step record.
 */
export type PilotPreviousStep = {
  /** Step description */
  step: string;
  /** Optional reviews */
  review?: PilotReview;
};
