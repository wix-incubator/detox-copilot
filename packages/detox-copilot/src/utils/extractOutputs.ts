export type OutputsMapping = Record<string, string>;

export const OUTPUTS_MAPPINGS: Record<string, OutputsMapping> = {
  PILOT_REVIEW_SECTION: {
    summary: "SUMMARY",
    findings: "FINDINGS",
    score: "SCORE",
  },
  PILOT_STEP: {
    thoughts: "THOUGHTS",
    action: "ACTION",
    ux: "UX",
    a11y: "ACCESSIBILITY",
  },
  PILOT_SUMMARY: {
    summary: "SUMMARY",
  },
};

export function extractOutputs<M extends OutputsMapping>({
  text,
  outputsMapper,
}: {
  text: string;
  outputsMapper: M;
}): { [K in keyof M]: string } {
  const outputs: Partial<{ [K in keyof M]: string }> = {};

  for (const fieldName in outputsMapper) {
    const tag = outputsMapper[fieldName];
    const regex = new RegExp(`<${tag}>(.*?)</${tag}>`, "s");
    const match = text.match(regex);
    if (match) {
      outputs[fieldName] = match[1].trim();
    } else {
      throw new Error(`Missing field for tag <${tag}>`);
    }
  }

  return outputs as { [K in keyof M]: string };
}
