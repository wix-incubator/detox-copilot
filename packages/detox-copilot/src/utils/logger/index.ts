import {
  createLogger,
  format,
  transports,
  Logger as WinstonLogger,
} from "winston";
import ora from "ora";
import chalk from "chalk";
import {
  LoggerMessageComponent,
  LoggerSpinner,
  LoggerMessageColor,
  LoggerOperationResultType,
} from "@/types/logger";

class Logger {
  private static instance: Logger;
  private readonly logger: WinstonLogger;
  private readonly logLevels = ["info", "warn", "error", "debug"] as const;
  private readonly colorMap: Record<
    (typeof this.logLevels)[number],
    LoggerMessageColor
  > = {
    info: "whiteBright",
    warn: "yellow",
    error: "red",
    debug: "gray",
  };

  private constructor() {
    this.logger = createLogger({
      level: "info",
      format: format.combine(
        format.colorize(),
        format.timestamp({ format: "YYYY-MM-DD HH:mm:ss" }),
        format.printf(
          ({ timestamp, level, message }) =>
            `[${timestamp}] ${level}: ${message}`,
        ),
      ),
      transports: [new transports.Console()],
    });
  }

  static getInstance(): Logger {
    if (!Logger.instance) {
      Logger.instance = new Logger();
    }
    return Logger.instance;
  }

  private colorizeMessage(...components: LoggerMessageComponent[]): string {
    return components
      .map((component) => {
        if (typeof component === "string") {
          return chalk.white(component);
        }

        let coloredMessage = chalk[component.color](component.message);

        if (component.isBold) {
          coloredMessage = chalk.bold(coloredMessage);
        }

        return coloredMessage;
      })
      .join(" ");
  }

  private log(
    level: (typeof this.logLevels)[number],
    ...components: LoggerMessageComponent[]
  ): void {
    const newComponents = components.map((component) => {
      if (typeof component === "string") {
        // Overriding the component to include the specified color
        return {
          message: component,
          isBold: false,
          color: this.colorMap[level],
        };
      }

      return component;
    });

    this.logger[level](this.colorizeMessage(...newComponents));
  }

  public info(...components: LoggerMessageComponent[]): void {
    this.log("info", ...components);
  }

  public warn(...components: LoggerMessageComponent[]): void {
    this.log("warn", ...components);
  }

  public error(...components: LoggerMessageComponent[]): void {
    this.log("error", ...components);
  }

  public debug(...components: LoggerMessageComponent[]): void {
    this.log("debug", ...components);
  }

  public startSpinner(...components: LoggerMessageComponent[]): LoggerSpinner {
    const spinner = ora(this.colorizeMessage(...components)).start();

    const stop = (
      result: LoggerOperationResultType,
      ...components: LoggerMessageComponent[]
    ) => {
      spinner.prefixText = "";
      const message = this.colorizeMessage(...components);

      const spinnerActions: Record<LoggerOperationResultType, () => void> = {
        success: () => spinner.succeed(message),
        failure: () => spinner.fail(message),
        warn: () => spinner.warn(message),
        info: () => spinner.info(message),
      };

      spinnerActions[result]!();
    };

    const update = (...components: LoggerMessageComponent[]) => {
      spinner.text = this.colorizeMessage(...components);
    };

    return { update, stop };
  }
}

export default Logger.getInstance();
