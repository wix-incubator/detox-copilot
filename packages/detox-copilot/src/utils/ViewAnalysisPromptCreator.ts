import { TestingFrameworkAPICatalog, PreviousStep } from "../types";

export class ViewAnalysisPromptCreator {
  constructor(private apiCatalog: TestingFrameworkAPICatalog) {}

  createPrompt(
    step: string,
    viewHierarchy: string,
    previousSteps: PreviousStep[] = [],
  ): string {
    return [
      "# View Hierarchy Analysis",
      "",
      "## Task Description",
      "",
      `Analyze the view hierarchy to identify the location and context of elements relevant to this task: "${step}"`,
      "",
      "## View Hierarchy",
      "```",
      viewHierarchy,
      "```",
      "",
      ...(previousSteps.length > 0
        ? [
            "## Previous Steps Context",
            "",
            ...previousSteps
              .map((previousStep, index) => [
                `### Step ${index + 1}`,
                `- Intent: "${previousStep.step}"`,
                `- Result: ${previousStep.result}`,
                "",
              ])
              .flat(),
          ]
        : []),
      "## Instructions",
      "",
      "1. Analyze the view hierarchy structure:",
      "   - Identify the target element's location",
      "   - Understand the surrounding layout context",
      "   - Note any parent-child relationships",
      "",
      "2. Evaluate element accessibility:",
      "   - Check if element is likely to be visible",
      "   - Identify potential scroll requirements",
      "   - Note any overlapping elements",
      "",
      "3. Describe the layout context:",
      "   - Parent container type (e.g., list, grid, navigation bar)",
      "   - Relative positioning (e.g., top of screen, within a card)",
      "   - Relationship to other UI elements",
      "",
      "Please provide your response in the following format:",
      "",
      "```",
      "Basic Element Description:",
      "A concise one-line description of what is being targeted (e.g., 'A button in a navigation bar', 'A text input in a form', 'The entire screen', 'A cell in a list')",
      "",
      "Element Location Analysis:",
      "1. Target Element Context",
      "   - Element Type: [The type of UI element being targeted]",
      "   - Container: [The parent container/view type]",
      "   - Position: [Where in the view hierarchy/screen the element is located]",
      "",
      "2. Accessibility Considerations",
      "   - Visibility: [Likely visibility status and any potential issues]",
      "   - Scroll Requirements: [Whether scrolling might be needed]",
      "   - Interaction Constraints: [Any potential barriers to interaction]",
      "",
      "3. Layout Structure",
      "   - Parent Hierarchy: [Description of the view hierarchy path to the element]",
      "   - Sibling Elements: [Related elements that might affect interaction]",
      "   - Screen Region: [General region of the screen where the element is located]",
      "```",
      "",
      "Additional Notes:",
      "- Focus on structural understanding rather than implementation details",
      "- Consider how the element's location affects test reliability",
      "- Note any potential challenges in locating or interacting with the element",
      "",
    ].join("\n");
  }
}
