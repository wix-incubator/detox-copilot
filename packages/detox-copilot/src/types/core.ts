import {
  TestingFrameworkDriver,
  TestingFrameworkAPICatalogCategory,
} from "@/types/framework";
import { PromptHandler } from "@/types/prompt";
import { PilotReport } from "@/types/pilot";

/**
 * Core interface for interacting with the Copilot testing automation system.
 */
export interface CopilotFacade {
  /**
   * Initializes the Copilot with the provided configuration.
   * Must be called once before using any other methods.
   * @param config - Configuration settings for the Copilot
   * @throws Error if called multiple times
   */
  init: (config: Config) => void;

  /**
   * Checks if Copilot has been properly initialized.
   * @returns true if initialized, false otherwise
   */
  isInitialized: () => boolean;

  /**
   * Starts a new test flow session.
   * Must be called before any test operations to ensure a clean state,
   * as Copilot uses operation history for context.
   */
  start: () => void;

  /**
   * Ends the current test flow session and handles cache management.
   * @param isCacheDisabled - If true, temporary cache data won't be persisted (default: false)
   */
  end: (isCacheDisabled?: boolean) => void;

  /**
   * Executes one or more test operations in sequence.
   * @param steps - Array of test step descriptions to perform
   * @returns Result of the final executed step
   * @example
   * // Single step
   * await perform('Tap on the login button')
   * // Multiple steps
   * await perform(
   *   'Scroll down to the 7th item',
   *   'The welcome message should be visible',
   *   'The welcome text should be "Hello"'
   * )
   */
  perform: (...steps: string[]) => Promise<string>;

  /**
   * Autonomously performs actions to achieve a specified testing goal.
   * @param goal - Description of the desired end state or objective
   * @returns Detailed report of actions taken and results
   * @example
   * await pilot('login with user "testuser" and password "pass123"')
   * await pilot('complete purchase flow in "my store"')
   */
  pilot: (goal: string) => Promise<PilotReport>;

  /**
   * Extends the testing framework's API capabilities.
   * @param categories - Additional API categories to add
   * @param context - Testing framework variables to expose
   */
  extendAPICatalog: (
    categories: TestingFrameworkAPICatalogCategory[],
    context?: any,
  ) => void;
}

/** Cache mode settings */
export type CacheMode = "full" | "lightweight" | "disabled";

/** Analysis mode settings */
export type AnalysisMode = "fast" | "full";

/**
 * Copilot behavior configuration options.
 */
export interface CopilotOptions {
  /** Cache mode (default: 'full') */
  cacheMode?: CacheMode;
  /** Analysis mode (default: 'fast') */
  analysisMode?: AnalysisMode;
}

/**
 * Complete Copilot configuration.
 */
export interface Config {
  /** Testing framework driver */
  frameworkDriver: TestingFrameworkDriver;
  /** AI service handler */
  promptHandler: PromptHandler;
  /** Optional behavior settings */
  options?: CopilotOptions;
}
