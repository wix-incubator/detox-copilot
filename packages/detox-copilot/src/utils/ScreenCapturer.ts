import { SnapshotManager } from "@/utils/SnapshotManager";
import { PromptHandler } from "@/types";
import logger from "@/utils/logger";

export class ScreenCapturer {
  constructor(
    private snapshotManager: SnapshotManager,
    private promptHandler: PromptHandler,
  ) {}

  async capture() {
    const loggerSpinner = logger.startSpinner(
      "Waiting for the screen to reach a stable state...",
    );

    const shouldCaptureSnapshot = this.promptHandler.isSnapshotImageSupported();
    const [snapshot, viewHierarchy] = await Promise.all([
      shouldCaptureSnapshot
        ? this.snapshotManager.captureSnapshotImage()
        : Promise.resolve(undefined),
      this.snapshotManager.captureViewHierarchyString(),
    ]);

    const isSnapshotImageAttached = snapshot != null && shouldCaptureSnapshot;

    loggerSpinner.stop(
      "success",
      "Screen has reached a stable state, captured the screen",
    );

    return { snapshot, viewHierarchy, isSnapshotImageAttached };
  }
}
