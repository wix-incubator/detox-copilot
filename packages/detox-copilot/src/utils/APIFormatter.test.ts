import { APIFormatter } from "@/utils/APIFormatter";
import { TestingFrameworkAPICatalog } from "@/types";

describe("APIFormatter", () => {
  const mockAPI: TestingFrameworkAPICatalog = {
    context: {},
    name: "Test Framework",
    description: "A testing framework for testing purposes",
    categories: [
      {
        title: "Actions",
        items: [
          {
            signature: "tap(element: Element)",
            description: "Taps on the specified element.",
            example: 'await element(by.id("button")).tap();',
            guidelines: [
              "Ensure the element is tappable before using this method.",
            ],
          },
          {
            signature: "typeText(element: Element, text: string)",
            description: "Types the specified text into the element.",
            example: 'await element(by.id("input")).typeText("Hello, World!");',
            guidelines: ["Use this method only on text input elements."],
          },
        ],
      },
      {
        title: "Assertions",
        items: [
          {
            signature: "toBeVisible()",
            description: "Asserts that the element is visible on the screen.",
            example: 'await expect(element(by.id("title"))).toBeVisible();',
            guidelines: ["Consider scroll position when using this assertion."],
          },
        ],
      },
      {
        title: "Matchers",
        items: [
          {
            signature: "by.id(id: string)",
            description: "Matches elements by their ID attribute.",
            example: 'element(by.id("uniqueId"))',
            guidelines: ["Use unique IDs for elements to avoid conflicts."],
          },
        ],
      },
    ],
  };

  let apiFormatter: APIFormatter;

  beforeEach(() => {
    apiFormatter = new APIFormatter(mockAPI);
  });

  describe("formatAPIMethod", () => {
    it("should format method with guidelines", () => {
      const method = mockAPI.categories[0].items[0];
      expect(apiFormatter.formatAPIMethod(method)).toMatchSnapshot();
    });

    it("should format method without guidelines", () => {
      const method = {
        ...mockAPI.categories[0].items[0],
        guidelines: undefined,
      };
      expect(apiFormatter.formatAPIMethod(method)).toMatchSnapshot();
    });
  });

  describe("formatAPICategory", () => {
    it("should format category with all its methods", () => {
      const category = mockAPI.categories[0];
      expect(apiFormatter.formatAPICategory(category)).toMatchSnapshot();
    });

    it("should format empty category", () => {
      const category = {
        title: "Empty Category",
        items: [],
      };
      expect(apiFormatter.formatAPICategory(category)).toMatchSnapshot();
    });
  });

  describe("formatAPIInfo", () => {
    it("should format all methods grouped by categories", () => {
      expect(apiFormatter.formatAPIInfo()).toMatchSnapshot();
    });

    it("should format empty API catalog", () => {
      const emptyFormatter = new APIFormatter({
        context: {},
        categories: [],
      });
      expect(emptyFormatter.formatAPIInfo()).toMatchSnapshot();
    });
  });
});
