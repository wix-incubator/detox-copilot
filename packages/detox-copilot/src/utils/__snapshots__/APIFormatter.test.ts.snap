// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`APIFormatter formatAPICategory should format category with all its methods 1`] = `
"### Actions

#### tap(element: Element)

Taps on the specified element.

##### Example

\`\`\`
await element(by.id("button")).tap();
\`\`\`

##### Guidelines

- Ensure the element is tappable before using this method.

#### typeText(element: Element, text: string)

Types the specified text into the element.

##### Example

\`\`\`
await element(by.id("input")).typeText("Hello, World!");
\`\`\`

##### Guidelines

- Use this method only on text input elements.
"
`;

exports[`APIFormatter formatAPICategory should format empty category 1`] = `
"### Empty Category
"
`;

exports[`APIFormatter formatAPIInfo should format all methods grouped by categories 1`] = `
"### Actions

#### tap(element: Element)

Taps on the specified element.

##### Example

\`\`\`
await element(by.id("button")).tap();
\`\`\`

##### Guidelines

- Ensure the element is tappable before using this method.

#### typeText(element: Element, text: string)

Types the specified text into the element.

##### Example

\`\`\`
await element(by.id("input")).typeText("Hello, World!");
\`\`\`

##### Guidelines

- Use this method only on text input elements.

### Assertions

#### toBeVisible()

Asserts that the element is visible on the screen.

##### Example

\`\`\`
await expect(element(by.id("title"))).toBeVisible();
\`\`\`

##### Guidelines

- Consider scroll position when using this assertion.

### Matchers

#### by.id(id: string)

Matches elements by their ID attribute.

##### Example

\`\`\`
element(by.id("uniqueId"))
\`\`\`

##### Guidelines

- Use unique IDs for elements to avoid conflicts.
"
`;

exports[`APIFormatter formatAPIInfo should format empty API catalog 1`] = `""`;

exports[`APIFormatter formatAPIMethod should format method with guidelines 1`] = `
"#### tap(element: Element)

Taps on the specified element.

##### Example

\`\`\`
await element(by.id("button")).tap();
\`\`\`

##### Guidelines

- Ensure the element is tappable before using this method.
"
`;

exports[`APIFormatter formatAPIMethod should format method without guidelines 1`] = `
"#### tap(element: Element)

Taps on the specified element.

##### Example

\`\`\`
await element(by.id("button")).tap();
\`\`\`
"
`;
